File: sdk/python/spiderweb_protocol/spiderweb_protocol/protocol.py
from __future__ import annotations

import asyncio
import base64
import json
from typing import Any, Protocol, TypedDict, cast

from .generated import (
    ACHERON_MESSAGE_TYPES,
    ACHERON_RUNTIME_VERSION,
    CONTROL_MESSAGE_TYPES,
    CONTROL_PROTOCOL,
    LEGACY_REJECTED_ACHERON_MESSAGE_TYPES,
    LEGACY_REJECTED_CONTROL_MESSAGE_TYPES,
    NODE_FS_PROTO,
    NODE_FS_PROTOCOL,
)


class SpiderProtocolError(RuntimeError):
    def __init__(self, code: str, message: str, details: Any | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.details = details


class ControlErrorPayload(TypedDict):
    code: str
    message: str


class AcheronErrorPayload(TypedDict, total=False):
    code: str
    errno: int
    message: str


class TextTransport(Protocol):
    async def send_text(self, text: str) -> None: ...
    async def receive_text(self) -> str: ...
    async def close(self) -> None: ...


class WebSocketTextTransport:
    def __init__(self, websocket: Any) -> None:
        self._websocket = websocket

    @classmethod
    async def connect(cls, url: str, **connect_kwargs: Any) -> "WebSocketTextTransport":
        try:
            import websockets
        except ImportError as exc:
            raise SpiderProtocolError(
                "missing_websocket_dependency",
                "install spiderweb-protocol[websocket] to use WebSocketTextTransport",
                exc,
            ) from exc

        websocket = await websockets.connect(url, **connect_kwargs)
        return cls(websocket)

    async def send_text(self, text: str) -> None:
        await self._websocket.send(text)

    async def receive_text(self) -> str:
        message = await self._websocket.recv()
        if not isinstance(message, str):
            raise SpiderProtocolError(
                "invalid_frame_type",
                "websocket transport expected a text frame",
                type(message).__name__,
            )
        return message

    async def close(self) -> None:
        await self._websocket.close()


def encode_data_b64(data: bytes | str) -> str:
    raw = data.encode("utf-8") if isinstance(data, str) else data
    return base64.b64encode(raw).decode("ascii")


def decode_data_b64(data_b64: str) -> bytes:
    return base64.b64decode(data_b64, validate=True)


def parse_envelope(raw: str | dict[str, Any]) -> dict[str, Any]:
    payload = _parse_json_object(raw) if isinstance(raw, str) else raw
    channel = _expect_string(payload, "channel")
    msg_type = _expect_string(payload, "type")

    if channel == "control":
        if not msg_type.startswith("control."):
            raise SpiderProtocolError(
                "namespace_mismatch",
                "control channel requires a control.* type",
                payload,
            )
        if msg_type in LEGACY_REJECTED_CONTROL_MESSAGE_TYPES:
            raise SpiderProtocolError(
                "unsupported_legacy_type",
                f"legacy control type {msg_type} is rejected",
                payload,
            )
        if msg_type not in CONTROL_MESSAGE_TYPES:
            raise SpiderProtocolError(
                "unsupported_type",
                f"unsupported control type {msg_type}",
                payload,
            )
        if "id" in payload and not isinstance(payload["id"], str):
            raise SpiderProtocolError("invalid_id", "control id must be a string", payload)
        return payload

    if channel == "acheron":
        if not msg_type.startswith("acheron."):
            raise SpiderProtocolError(
                "namespace_mismatch",
                "acheron channel requires an acheron.* type",
                payload,
            )
        if msg_type in LEGACY_REJECTED_ACHERON_MESSAGE_TYPES:
            raise SpiderProtocolError(
                "unsupported_legacy_type",
                f"legacy acheron type {msg_type} is rejected",
                payload,
            )
        if msg_type not in ACHERON_MESSAGE_TYPES:
            raise SpiderProtocolError(
                "unsupported_type",
                f"unsupported acheron type {msg_type}",
                payload,
            )
        if "tag" in payload and (not isinstance(payload["tag"], int) or isinstance(payload["tag"], bool)):
            raise SpiderProtocolError("invalid_tag", "acheron tag must be an integer", payload)
        return payload

    raise SpiderProtocolError("invalid_channel", f"unsupported channel {channel}", payload)


def stringify_envelope(envelope: dict[str, Any]) -> str:
    return json.dumps(envelope, separators=(",", ":"))


def build_control_envelope(
    msg_type: str,
    *,
    request_id: str | None = None,
    ok: bool | None = None,
    payload: Any | None = None,
    error: ControlErrorPayload | None = None,
) -> dict[str, Any]:
    envelope: dict[str, Any] = {"channel": "control", "type": msg_type}
    if request_id is not None:
        envelope["id"] = request_id
    if ok is not None:
        envelope["ok"] = ok
    if payload is not None:
        envelope["payload"] = payload
    if error is not None:
        envelope["error"] = error
    return envelope


def build_control_error(code: str, message: str, request_id: str | None = None) -> dict[str, Any]:
    return build_control_envelope(
        "control.error",
        request_id=request_id,
        ok=False,
        error={"code": code, "message": message},
    )


def build_control_version_request(request_id: str) -> dict[str, Any]:
    return build_control_envelope(
        "control.version",
        request_id=request_id,
        payload={"protocol": CONTROL_PROTOCOL},
    )


def build_control_connect_request(request_id: str) -> dict[str, Any]:
    return build_control_envelope("control.connect", request_id=request_id, payload={})


def build_acheron_envelope(
    msg_type: str,
    *,
    tag: int | None = None,
    ok: bool | None = None,
    payload: Any | None = None,
    error: AcheronErrorPayload | None = None,
    **extra: Any,
) -> dict[str, Any]:
    envelope: dict[str, Any] = {"channel": "acheron", "type": msg_type}
    if tag is not None:
        envelope["tag"] = tag
    if ok is not None:
        envelope["ok"] = ok
    if payload is not None:
        envelope["payload"] = payload
    if error is not None:
        envelope["error"] = error
    envelope.update(extra)
    return envelope


def build_acheron_error(code: str, message: str, tag: int | None = None) -> dict[str, Any]:
    return build_acheron_envelope(
        "acheron.error",
        tag=tag,
        ok=False,
        error={"code": code, "message": message},
    )


def build_acheron_fs_error(errno: int, message: str, tag: int | None = None) -> dict[str, Any]:
    return build_acheron_envelope(
        "acheron.err_fs",
        tag=tag,
        ok=False,
        error={"errno": errno, "message": message},
    )


def build_acheron_version_request(tag: int, msize: int = 1_048_576) -> dict[str, Any]:
    return build_acheron_envelope(
        "acheron.t_version",
        tag=tag,
        msize=msize,
        version=ACHERON_RUNTIME_VERSION,
    )


def build_acheron_attach_request(tag: int, fid: int = 1) -> dict[str, Any]:
    return build_acheron_envelope("acheron.t_attach", tag=tag, fid=fid)


def build_fs_hello_request(tag: int) -> dict[str, Any]:
    return build_acheron_envelope(
        "acheron.t_fs_hello",
        tag=tag,
        payload={"protocol": NODE_FS_PROTOCOL, "proto": NODE_FS_PROTO},
    )


class ControlClient:
    def __init__(self, transport: TextTransport) -> None:
        self._transport = transport

    async def negotiate_version(self, request_id: str = "control-version") -> dict[str, Any]:
        return await self.request("control.version", request_id, {"protocol": CONTROL_PROTOCOL})

    async def connect(self, request_id: str = "control-connect") -> dict[str, Any]:
        return await self.request("control.connect", request_id, {})

    async def request(self, msg_type: str, request_id: str, payload: Any | None = None) -> dict[str, Any]:
        await self._transport.send_text(
            stringify_envelope(build_control_envelope(msg_type, request_id=request_id, payload=payload))
        )
        while True:
            envelope = parse_envelope(await self._transport.receive_text())
            if envelope["channel"] != "control":
                continue
            if envelope.get("id") != request_id:
                continue
            if envelope["type"] == "control.error":
                raise _protocol_error_from_control(envelope)
            return envelope


class AcheronClient:
    def __init__(self, transport: TextTransport) -> None:
        self._transport = transport

    async def negotiate_version(self, tag: int = 1, msize: int = 1_048_576) -> dict[str, Any]:
        return await self.request(build_acheron_version_request(tag, msize), "acheron.r_version")

    async def attach(self, tag: int = 2, fid: int = 1) -> dict[str, Any]:
        return await self.request(build_acheron_attach_request(tag, fid), "acheron.r_attach")

    async def request(
        self,
        envelope: dict[str, Any],
        expected_type: str,
        *,
        on_event: Any | None = None,
    ) -> dict[str, Any]:
        await self._transport.send_text(stringify_envelope(envelope))
        while True:
            parsed = parse_envelope(await self._transport.receive_text())
            if parsed["channel"] != "acheron":
                continue
            if _is_acheron_event(parsed):
                if on_event is not None:
                    await _maybe_await(on_event(parsed))
                continue
            if parsed.get("tag") != envelope.get("tag"):
                continue
            if parsed["type"] in {"acheron.error", "acheron.err_fs"}:
                raise _protocol_error_from_acheron(parsed)
            if parsed["type"] != expected_type:
                raise SpiderProtocolError(
                    "unexpected_type",
                    f"expected {expected_type} but received {parsed['type']}",
                    parsed,
                )
            return parsed


class FsClient:
    def __init__(self, transport: TextTransport) -> None:
        self._acheron = AcheronClient(transport)

    async def hello(self, tag: int = 1) -> dict[str, Any]:
        return await self._acheron.request(build_fs_hello_request(tag), "acheron.r_fs_hello")

    async def lookup(self, tag: int, node: int, name: str) -> dict[str, Any]:
        return await self._request("acheron.t_fs_lookup", "acheron.r_fs_lookup", tag, node=node, payload={"name": name})

    async def getattr(self, tag: int, node: int) -> dict[str, Any]:
        return await self._request("acheron.t_fs_getattr", "acheron.r_fs_getattr", tag, node=node, payload={})

    async def readdirp(self, tag: int, node: int, cookie: int = 0, count: int = 128) -> dict[str, Any]:
        return await self._request(
            "acheron.t_fs_readdirp",
            "acheron.r_fs_readdirp",
            tag,
            node=node,
            payload={"cookie": cookie, "count": count},
        )

    async def open(self, tag: int, node: int, mode: str = "r") -> dict[str, Any]:
        return await self._request("acheron.t_fs_open", "acheron.r_fs_open", tag, node=node, payload={"mode": mode})

    async def read(self, tag: int, handle: int, offset: int = 0, count: int = 4096) -> dict[str, Any]:
        return await self._request(
            "acheron.t_fs_read",
            "acheron.r_fs_read",
            tag,
            h=handle,
            payload={"offset": offset, "count": count},
        )

    async def write(self, tag: int, handle: int, data: bytes | str, offset: int = 0) -> dict[str, Any]:
        return await self._request(
            "acheron.t_fs_write",
            "acheron.r_fs_write",
            tag,
            h=handle,
            payload={"offset": offset, "data_b64": encode_data_b64(data)},
        )

    async def close(self, tag: int, handle: int) -> dict[str, Any]:
        return await self._request("acheron.t_fs_close", "acheron.r_fs_close", tag, h=handle, payload={})

    async def _request(self, msg_type: str, expected_type: str, tag: int, **extra: Any) -> dict[str, Any]:
        envelope = build_acheron_envelope(msg_type, tag=tag, **extra)
        return await self._acheron.request(envelope, expected_type)


def _parse_json_object(raw: str) -> dict[str, Any]:
    try:
        value = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise SpiderProtocolError("invalid_json", "message is not valid JSON", exc) from exc
    if not isinstance(value, dict):
        raise SpiderProtocolError("invalid_envelope", "message must decode to an object", value)
    return cast(dict[str, Any], value)


def _expect_string(payload: dict[str, Any], key: str) -> str:
    value = payload.get(key)
    if not isinstance(value, str):
        raise SpiderProtocolError("missing_field", f"{key} must be a string", payload)
    return value


def _protocol_error_from_control(envelope: dict[str, Any]) -> SpiderProtocolError:
    error = envelope.get("error")
    if not isinstance(error, dict):
        return SpiderProtocolError("control_error", "control request failed", envelope)
    return SpiderProtocolError(
        cast(str, error.get("code", "control_error")),
        cast(str, error.get("message", "control request failed")),
        envelope,
    )


def _protocol_error_from_acheron(envelope: dict[str, Any]) -> SpiderProtocolError:
    error = envelope.get("error")
    if not isinstance(error, dict):
        return SpiderProtocolError("acheron_error", "acheron request failed", envelope)
    if isinstance(error.get("errno"), int):
        return SpiderProtocolError(
            "acheron_fs_error",
            cast(str, error.get("message", "acheron fs request failed")),
            envelope,
        )
    return SpiderProtocolError(
        cast(str, error.get("code", "acheron_error")),
        cast(str, error.get("message", "acheron request failed")),
        envelope,
    )


def _is_acheron_event(envelope: dict[str, Any]) -> bool:
    return envelope.get("type") in {"acheron.e_fs_inval", "acheron.e_fs_inval_dir"}


async def _maybe_await(value: Any) -> Any:
    if asyncio.iscoroutine(value):
        return await value
    return value

File: sdk/rust/spiderweb-protocol/src/protocol.rs
use std::collections::VecDeque;

use async_trait::async_trait;
use base64::Engine as _;
use serde_json::{Map, Value};
use thiserror::Error;

use crate::generated::*;

const LEGACY_REJECTED_CONTROL_MESSAGE_TYPES: &[&str] = &[
    "session.send",
    "control.debug_subscribe",
    "control.debug_unsubscribe",
    "control.node_service_watch",
    "control.node_service_unwatch",
    "control.node_service_event",
    "control.venom_watch",
    "control.venom_unwatch",
    "control.venom_event",
];

const LEGACY_REJECTED_ACHERON_MESSAGE_TYPES: &[&str] = &[
    "acheron.t_hello",
    "acheron.fs_t_*",
    "acheron.fs_r_*",
    "acheron.fs_evt_*",
    "acheron.fs_error",
];

#[derive(Debug, Clone, Error)]
#[error("{message}")]
pub struct SpiderProtocolError {
    pub code: String,
    pub message: String,
    pub details: Option<Value>,
}

impl SpiderProtocolError {
    pub fn new(code: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            code: code.into(),
            message: message.into(),
            details: None,
        }
    }

    pub fn with_details(mut self, details: Value) -> Self {
        self.details = Some(details);
        self
    }
}

#[async_trait]
pub trait TextTransport: Send {
    async fn send_text(&mut self, text: String) -> Result<(), SpiderProtocolError>;
    async fn receive_text(&mut self) -> Result<String, SpiderProtocolError>;
    async fn close(&mut self) -> Result<(), SpiderProtocolError>;
}

pub fn encode_data_b64(data: impl AsRef<[u8]>) -> String {
    base64::engine::general_purpose::STANDARD.encode(data)
}

pub fn decode_data_b64(data_b64: &str) -> Result<Vec<u8>, SpiderProtocolError> {
    base64::engine::general_purpose::STANDARD
        .decode(data_b64)
        .map_err(|error| SpiderProtocolError::new("invalid_base64", "base64 decode failed").with_details(Value::String(error.to_string())))
}

pub fn stringify_control_request(envelope: &ControlRequestEnvelope) -> Result<String, SpiderProtocolError> {
    stringify_value(envelope.to_value())
}

pub fn stringify_control_response(envelope: &ControlResponseEnvelope) -> Result<String, SpiderProtocolError> {
    stringify_value(envelope.to_value())
}

pub fn stringify_control_error(envelope: &ControlErrorEnvelope) -> Result<String, SpiderProtocolError> {
    stringify_value(serde_json::to_value(envelope))
}

pub fn stringify_acheron_request(envelope: &AcheronRequestEnvelope) -> Result<String, SpiderProtocolError> {
    stringify_value(envelope.to_value())
}

pub fn stringify_acheron_response(envelope: &AcheronResponseEnvelope) -> Result<String, SpiderProtocolError> {
    stringify_value(envelope.to_value())
}

pub fn stringify_acheron_event(envelope: &AcheronEventEnvelopeEnum) -> Result<String, SpiderProtocolError> {
    stringify_value(envelope.to_value())
}

pub fn stringify_acheron_error(envelope: &AcheronErrorEnvelopeEnum) -> Result<String, SpiderProtocolError> {
    stringify_value(envelope.to_value())
}

pub fn parse_control_request(raw: &str) -> Result<ControlRequestEnvelope, SpiderProtocolError> {
    let value = parse_raw_object(raw)?;
    validate_control_value(&value)?;
    ControlRequestEnvelope::from_value(value).map_err(json_error)
}

pub fn parse_control_response(raw: &str) -> Result<ControlResponseEnvelope, SpiderProtocolError> {
    let value = parse_raw_object(raw)?;
    validate_control_value(&value)?;
    if value.get("type").and_then(Value::as_str) == Some("control.error") {
        return Err(protocol_error_from_control_value(&value));
    }
    ControlResponseEnvelope::from_value(value).map_err(json_error)
}

pub fn parse_control_error(raw: &str) -> Result<ControlErrorEnvelope, SpiderProtocolError> {
    let value = parse_raw_object(raw)?;
    validate_control_value(&value)?;
    serde_json::from_value(value).map_err(json_error)
}

pub fn parse_acheron_request(raw: &str) -> Result<AcheronRequestEnvelope, SpiderProtocolError> {
    let value = parse_raw_object(raw)?;
    validate_acheron_value(&value)?;
    AcheronRequestEnvelope::from_value(value).map_err(json_error)
}

pub fn parse_acheron_response(raw: &str) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
    let value = parse_raw_object(raw)?;
    validate_acheron_value(&value)?;
    let message_type = value.get("type").and_then(Value::as_str).unwrap_or_default();
    if message_type == "acheron.error" || message_type == "acheron.err_fs" {
        return Err(protocol_error_from_acheron_value(&value));
    }
    AcheronResponseEnvelope::from_value(value).map_err(json_error)
}

pub fn parse_acheron_event(raw: &str) -> Result<AcheronEventEnvelopeEnum, SpiderProtocolError> {
    let value = parse_raw_object(raw)?;
    validate_acheron_value(&value)?;
    AcheronEventEnvelopeEnum::from_value(value).map_err(json_error)
}

pub fn parse_acheron_error(raw: &str) -> Result<AcheronErrorEnvelopeEnum, SpiderProtocolError> {
    let value = parse_raw_object(raw)?;
    validate_acheron_value(&value)?;
    AcheronErrorEnvelopeEnum::from_value(value).map_err(json_error)
}

pub struct ControlClient<T: TextTransport> {
    transport: T,
}

impl<T: TextTransport> ControlClient<T> {
    pub fn new(transport: T) -> Self {
        Self { transport }
    }

    pub fn into_inner(self) -> T {
        self.transport
    }

    pub async fn negotiate_version(&mut self, request_id: impl Into<String>) -> Result<ControlResponseEnvelope, SpiderProtocolError> {
        let request_id = request_id.into();
        self.request(ControlRequestEnvelope::Version(ControlEnvelope {
            channel: Channel::Control,
            message_type: ControlMessageType::Version,
            id: Some(request_id),
            ok: None,
            payload: Some(ControlVersionRequestPayload {
                protocol: CONTROL_PROTOCOL.to_string(),
            }),
        }))
        .await
    }

    pub async fn connect(&mut self, request_id: impl Into<String>) -> Result<ControlResponseEnvelope, SpiderProtocolError> {
        self.request(ControlRequestEnvelope::Connect(ControlEnvelope {
            channel: Channel::Control,
            message_type: ControlMessageType::Connect,
            id: Some(request_id.into()),
            ok: None,
            payload: Some(EmptyObject::default()),
        }))
        .await
    }

    pub async fn request(&mut self, request: ControlRequestEnvelope) -> Result<ControlResponseEnvelope, SpiderProtocolError> {
        let request_id = control_request_id(&request)?;
        self.transport.send_text(stringify_control_request(&request)?).await?;
        loop {
            let raw = self.transport.receive_text().await?;
            let value = parse_raw_object(&raw)?;
            if value.get("channel").and_then(Value::as_str) != Some("control") {
                continue;
            }
            if value.get("id").and_then(Value::as_str) != Some(request_id.as_str()) {
                continue;
            }
            if value.get("type").and_then(Value::as_str) == Some("control.error") {
                return Err(protocol_error_from_control_value(&value));
            }
            validate_control_value(&value)?;
            return ControlResponseEnvelope::from_value(value).map_err(json_error);
        }
    }
}

pub struct AcheronClient<T: TextTransport> {
    transport: T,
}

impl<T: TextTransport> AcheronClient<T> {
    pub fn new(transport: T) -> Self {
        Self { transport }
    }

    pub fn into_inner(self) -> T {
        self.transport
    }

    pub async fn negotiate_version(&mut self, tag: u32, msize: u32) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.request(AcheronRequestEnvelope::TVersion(AcheronVersionEnvelope {
            channel: Channel::Acheron,
            message_type: AcheronMessageType::TVersion,
            tag: Some(tag),
            ok: None,
            msize,
            version: ACHERON_RUNTIME_VERSION.to_string(),
        }))
        .await
    }

    pub async fn attach(&mut self, tag: u32, fid: u32) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.request(AcheronRequestEnvelope::TAttach(AcheronAttachEnvelope {
            channel: Channel::Acheron,
            message_type: AcheronMessageType::TAttach,
            tag: Some(tag),
            ok: None,
            fid,
        }))
        .await
    }

    pub async fn request(&mut self, request: AcheronRequestEnvelope) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.request_with_events(request, |_| Ok(())).await
    }

    pub async fn request_with_events<F>(&mut self, request: AcheronRequestEnvelope, mut on_event: F) -> Result<AcheronResponseEnvelope, SpiderProtocolError>
    where
        F: FnMut(AcheronEventEnvelopeEnum) -> Result<(), SpiderProtocolError>,
    {
        let request_tag = acheron_request_tag(&request)?;
        self.transport.send_text(stringify_acheron_request(&request)?).await?;
        loop {
            let raw = self.transport.receive_text().await?;
            let value = parse_raw_object(&raw)?;
            if value.get("channel").and_then(Value::as_str) != Some("acheron") {
                continue;
            }
            validate_acheron_value(&value)?;
            let message_type = value.get("type").and_then(Value::as_str).unwrap_or_default();
            if message_type == "acheron.e_fs_inval" || message_type == "acheron.e_fs_inval_dir" {
                on_event(AcheronEventEnvelopeEnum::from_value(value).map_err(json_error)?)?;
                continue;
            }
            if value.get("tag").and_then(Value::as_u64) != Some(u64::from(request_tag)) {
                continue;
            }
            if message_type == "acheron.error" || message_type == "acheron.err_fs" {
                return Err(protocol_error_from_acheron_value(&value));
            }
            return AcheronResponseEnvelope::from_value(value).map_err(json_error);
        }
    }
}

pub struct FsClient<T: TextTransport> {
    acheron: AcheronClient<T>,
}

impl<T: TextTransport> FsClient<T> {
    pub fn new(transport: T) -> Self {
        Self {
            acheron: AcheronClient::new(transport),
        }
    }

    pub fn into_inner(self) -> T {
        self.acheron.into_inner()
    }

    pub async fn hello(&mut self, tag: u32) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTHello(AcheronPayloadEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTHello,
                tag: Some(tag),
                ok: None,
                payload: Some(FsHelloRequest {
                    protocol: NODE_FS_PROTOCOL.to_string(),
                    proto: NODE_FS_PROTO,
                    auth_token: None,
                    node_id: None,
                    node_secret: None,
                }),
            }))
            .await
    }

    pub async fn lookup(&mut self, tag: u32, node: u64, name: impl Into<String>) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTLookup(AcheronNodeEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTLookup,
                tag: Some(tag),
                ok: None,
                node,
                payload: Some(FsLookupRequest { name: name.into() }),
            }))
            .await
    }

    pub async fn getattr(&mut self, tag: u32, node: u64) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTGetattr(AcheronNodeEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTGetattr,
                tag: Some(tag),
                ok: None,
                node,
                payload: Some(EmptyObject::default()),
            }))
            .await
    }

    pub async fn readdirp(&mut self, tag: u32, node: u64, cookie: u64, count: u32) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTReaddirp(AcheronNodeEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTReaddirp,
                tag: Some(tag),
                ok: None,
                node,
                payload: Some(FsReaddirpRequest {
                    cookie: Some(cookie),
                    count: Some(count),
                    max: None,
                }),
            }))
            .await
    }

    pub async fn open(&mut self, tag: u32, node: u64, mode: impl Into<String>) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTOpen(AcheronNodeEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTOpen,
                tag: Some(tag),
                ok: None,
                node,
                payload: Some(FsOpenRequest {
                    mode: Some(mode.into()),
                    flags: None,
                }),
            }))
            .await
    }

    pub async fn read(&mut self, tag: u32, handle: u64, offset: u64, count: u32) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTRead(AcheronHandleEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTRead,
                tag: Some(tag),
                ok: None,
                handle,
                payload: Some(FsReadRequest {
                    offset,
                    count,
                }),
            }))
            .await
    }

    pub async fn write(&mut self, tag: u32, handle: u64, offset: u64, data: impl AsRef<[u8]>) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTWrite(AcheronHandleEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTWrite,
                tag: Some(tag),
                ok: None,
                handle,
                payload: Some(FsWriteRequest {
                    offset,
                    data_b64: encode_data_b64(data),
                }),
            }))
            .await
    }

    pub async fn close(&mut self, tag: u32, handle: u64) -> Result<AcheronResponseEnvelope, SpiderProtocolError> {
        self.acheron
            .request(AcheronRequestEnvelope::FsTClose(AcheronHandleEnvelope {
                channel: Channel::Acheron,
                message_type: AcheronMessageType::FsTClose,
                tag: Some(tag),
                ok: None,
                handle,
                payload: Some(EmptyObject::default()),
            }))
            .await
    }
}

fn parse_raw_object(raw: &str) -> Result<Value, SpiderProtocolError> {
    let value: Value = serde_json::from_str(raw).map_err(json_error)?;
    if !value.is_object() {
        return Err(SpiderProtocolError::new("invalid_envelope", "message must decode to an object"));
    }
    Ok(value)
}

fn validate_control_value(value: &Value) -> Result<(), SpiderProtocolError> {
    let channel = expect_string(value, "channel")?;
    if channel != "control" {
        return Err(SpiderProtocolError::new("invalid_channel", "unsupported control channel").with_details(value.clone()));
    }
    let message_type = expect_string(value, "type")?;
    if !message_type.starts_with("control.") {
        return Err(SpiderProtocolError::new("namespace_mismatch", "control channel requires a control.* type").with_details(value.clone()));
    }
    if LEGACY_REJECTED_CONTROL_MESSAGE_TYPES.contains(&message_type) {
        return Err(SpiderProtocolError::new("unsupported_legacy_type", "legacy control type is rejected").with_details(value.clone()));
    }
    if let Some(id) = value.get("id") {
        if !id.is_string() {
            return Err(SpiderProtocolError::new("invalid_id", "control id must be a string").with_details(value.clone()));
        }
    }
    Ok(())
}

fn validate_acheron_value(value: &Value) -> Result<(), SpiderProtocolError> {
    let channel = expect_string(value, "channel")?;
    if channel != "acheron" {
        return Err(SpiderProtocolError::new("invalid_channel", "unsupported acheron channel").with_details(value.clone()));
    }
    let message_type = expect_string(value, "type")?;
    if !message_type.starts_with("acheron.") {
        return Err(SpiderProtocolError::new("namespace_mismatch", "acheron channel requires an acheron.* type").with_details(value.clone()));
    }
    if LEGACY_REJECTED_ACHERON_MESSAGE_TYPES.contains(&message_type) {
        return Err(SpiderProtocolError::new("unsupported_legacy_type", "legacy acheron type is rejected").with_details(value.clone()));
    }
    if let Some(tag) = value.get("tag") {
        if tag.as_u64().is_none() {
            return Err(SpiderProtocolError::new("invalid_tag", "acheron tag must be an integer").with_details(value.clone()));
        }
    }
    Ok(())
}

fn expect_string<'a>(value: &'a Value, key: &str) -> Result<&'a str, SpiderProtocolError> {
    value
        .get(key)
        .and_then(Value::as_str)
        .ok_or_else(|| SpiderProtocolError::new("missing_field", format!("{key} must be a string")).with_details(value.clone()))
}

fn stringify_value(value: serde_json::Result<Value>) -> Result<String, SpiderProtocolError> {
    let value = value.map_err(json_error)?;
    serde_json::to_string(&value).map_err(json_error)
}

fn json_error(error: serde_json::Error) -> SpiderProtocolError {
    SpiderProtocolError::new("invalid_json", "json serialization or parse failed").with_details(Value::String(error.to_string()))
}

fn control_request_id(request: &ControlRequestEnvelope) -> Result<String, SpiderProtocolError> {
    request
        .to_value()
        .map_err(json_error)?
        .get("id")
        .and_then(Value::as_str)
        .map(ToOwned::to_owned)
        .ok_or_else(|| SpiderProtocolError::new("missing_id", "control request id is required"))
}

fn acheron_request_tag(request: &AcheronRequestEnvelope) -> Result<u32, SpiderProtocolError> {
    request
        .to_value()
        .map_err(json_error)?
        .get("tag")
        .and_then(Value::as_u64)
        .and_then(|value| u32::try_from(value).ok())
        .ok_or_else(|| SpiderProtocolError::new("missing_tag", "acheron request tag is required"))
}

fn protocol_error_from_control_value(value: &Value) -> SpiderProtocolError {
    let error = value.get("error").and_then(Value::as_object);
    let code = error
        .and_then(|inner| inner.get("code"))
        .and_then(Value::as_str)
        .unwrap_or("control_error");
    let message = error
        .and_then(|inner| inner.get("message"))
        .and_then(Value::as_str)
        .unwrap_or("control request failed");
    SpiderProtocolError::new(code, message).with_details(value.clone())
}

fn protocol_error_from_acheron_value(value: &Value) -> SpiderProtocolError {
    let error = value.get("error").and_then(Value::as_object);
    if let Some(errno) = error.and_then(|inner| inner.get("errno")).and_then(Value::as_u64) {
        return SpiderProtocolError::new(
            "acheron_fs_error",
            error
                .and_then(|inner| inner.get("message"))
                .and_then(Value::as_str)
                .unwrap_or("acheron fs request failed"),
        )
        .with_details(Value::Object(Map::from_iter([
            ("envelope".to_string(), value.clone()),
            ("errno".to_string(), Value::from(errno)),
        ])));
    }
    SpiderProtocolError::new(
        error
            .and_then(|inner| inner.get("code"))
            .and_then(Value::as_str)
            .unwrap_or("acheron_error"),
        error
            .and_then(|inner| inner.get("message"))
            .and_then(Value::as_str)
            .unwrap_or("acheron request failed"),
    )
    .with_details(value.clone())
}

pub struct MockTextTransport {
    sent: Vec<String>,
    incoming: VecDeque<String>,
}

impl MockTextTransport {
    pub fn new(incoming: impl IntoIterator<Item = String>) -> Self {
        Self {
            sent: Vec::new(),
            incoming: incoming.into_iter().collect(),
        }
    }

    pub fn sent(&self) -> &[String] {
        &self.sent
    }
}

#[async_trait]
impl TextTransport for MockTextTransport {
    async fn send_text(&mut self, text: String) -> Result<(), SpiderProtocolError> {
        self.sent.push(text);
        Ok(())
    }

    async fn receive_text(&mut self) -> Result<String, SpiderProtocolError> {
        self.incoming
            .pop_front()
            .ok_or_else(|| SpiderProtocolError::new("transport_eof", "mock transport is out of messages"))
    }

    async fn close(&mut self) -> Result<(), SpiderProtocolError> {
        Ok(())
    }
}
